### ISSUE TITLE
Confusion about whitelist_weight_modules and blacklist_weight_modules

### ISSUE THREAD

**Issue by mc451742** [fudan-zvg/RealMotion#2]
Dear authors, apologize to disturb you again.

In line 168 of your 'pl_modules.py' source code file, there are two parameters 'whitelist_weight_modules' and 'blacklist_weight_modules'. I am quite confused about the purposes of these two parameters.

The source code is here, starting from line 165:

```

def configure_optimizers(self):
        decay = set()
        no_decay = set()
        whitelist_weight_modules = (
            nn.Linear,
            nn.Conv1d,
            nn.Conv2d,
            nn.Conv3d,
            nn.MultiheadAttention,
            nn.LSTM,
            nn.GRU,
        )
        blacklist_weight_modules = (
            nn.BatchNorm1d,
            nn.BatchNorm2d,
            nn.BatchNorm3d,
            nn.SyncBatchNorm,
            nn.LayerNorm,
            nn.Embedding,
        )
        for module_name, module in self.named_modules():
            for param_name, param in module.named_parameters():
                full_param_name = (
                    '%s.%s' % (module_name, param_name) if module_name else param_name
                )
                if 'bias' in param_name:
                    no_decay.add(full_param_name)
                elif 'weight' in param_name:
                    if isinstance(module, whitelist_weight_modules):
                        decay.add(full_param_name)
                    elif isinstance(module, blacklist_weight_modules):
                        no_decay.add(full_param_name)
                elif not ('weight' in param_name or 'bias' in param_name):
                    no_decay.add(full_param_name)
        param_dict = {
            param_name: param for param_name, param in self.named_parameters()
        }
        inter_params = decay & no_decay
        union_params = decay | no_decay
        assert len(inter_params) == 0
        assert len(param_dict.keys() - union_params) == 0

        optim_groups = [
            {
                'params': [
                    param_dict[param_name] for param_name in sorted(list(decay))
                ],
                'weight_decay': self.optim.weight_decay,
            },
            {
                'params': [
                    param_dict[param_name] for param_name in sorted(list(no_decay))
                ],
                'weight_decay': 0.0,
            },
        ]

        optimizer = torch.optim.AdamW(
            optim_groups, lr=self.optim.lr, weight_decay=self.optim.weight_decay
        )
        scheduler = WarmupCosLR(
            optimizer=optimizer,
            lr=self.optim.lr,
            min_lr=self.optim.min_lr,
            warmup_ratio=self.optim.warmup_ratio,
            epochs=self.optim.epochs,
        )
        return [optimizer], [scheduler]

```

It looks like in 'decay' and 'no_decay', the two parameters were used. I am just quite confused about this part. What are the purposes of using them and why it is used in 'decay' and 'no_decay'? What are the functions of 'decay' and 'no_decay'?

Sorry for this issue and please bear with me. I am looking forward for your reply. Much appreciate it.

**Comment by mc451742** [comment]
Oh sorry for the questions proposed about 'decay' and 'no_decay'. I think I figure it out myself. But I am still confused about the  'whitelist_weight_modules' and 'blacklist_weight_modules'.

**Comment by NaNaoiSong** [comment]
This code is migrated from another repository and it follows the common practice that bias and norm layers are less related to overfitting. Hence, weight decay is only applied to the weights of the modules listed in 'whitelist_weight_modules'. You can try to remove 'optim_groups', as we think it would have minimal impact on performance.

**Comment by mc451742** [comment]
Got it. Much appreciate for your reply and explain! That's very helpful!
